packages/hardhat-core/src/internal/cli/autocomplete.ts

```diff
@@ -62,50 +95,226 @@ export async function complete({
   }
 
   if (prev === "--network") {
-    return Object.keys(hre.config.networks);
+    return networks.filter(startsWithLast);
   }
 
   // if the previous word is a param, then a value is expected
   // we don't complete anything here
   if (prev.startsWith("-")) {
     const paramName = ArgumentsParser.cLAToParamName(prev);
 
-    const globalParam: any = (HARDHAT_PARAM_DEFINITIONS as any)[paramName];
+    const globalParam = HARDHAT_PARAM_DEFINITIONS[paramName as GlobalParam];
     if (globalParam !== undefined && !globalParam.isFlag) {
-      return [];
+      return filesystemSuggestions(last);
     }
   }
 
   // if there's no task, we complete either tasks or params
-  if (task === undefined || hre.tasks[task] === undefined) {
-    const tasks = Object.values(hre.tasks)
+  if (task === undefined || tasks[task] === undefined) {
+    const taskNames = Object.values(tasks)
       .map((x) => x.name)
       .filter((x) => !x.includes(":"));
     if (last.startsWith("-")) {
-      return coreParams;
+      return coreParams.filter(startsWithLast);
     }
-    return tasks;
+    return taskNames.filter(startsWithLast);
   }
 
   if (!last.startsWith("-")) {
-    return [];
+    return filesystemSuggestions(last);
   }
 
   // if there's a task and the last word starts with -, we complete its params and the global params
-  const taskParams = Object.values(hre.tasks[task].paramDefinitions)
+  const taskParams = Object.values(tasks[task].paramDefinitions)
     .map((x) => x.name)
     .map(ArgumentsParser.paramNameToCLA)
     .filter((x) => !words.includes(x));
 
-  return [...taskParams, ...coreParams];
+  return [...taskParams, ...coreParams].filter((completion) =>
+    completion.startsWith(last)
+  );
+}
+
+async function getCompletionData(): Promise<CompletionData | undefined> {
+  const projectName = getProjectName();
+  if (projectName === undefined) {
+    return undefined;
+  }
+
+  const cachedCompletionDataPerProject = await getCachedCompletionDataPerProject();
+  const cachedCompletionData = cachedCompletionDataPerProject?.[projectName];
+
+  if (cachedCompletionData !== undefined) {
+    if (arePreviousMtimesCorrect(cachedCompletionData.mtimes)) {
+      return cachedCompletionData.completionData;
+    }
+  }
+
+  const filesBeforeRequire = Object.keys(require.cache);
+  let hre: typeof hardhat;
+  try {
+    process.env.TS_NODE_TRANSPILE_ONLY = "1";
+    require("../../register");
```

Comment:
I still don't understand why this is necessary. There's a double initialization going on here, which makes things slower.


What's the error you get if you do `hre = require("../..")`?

Reply:
```
SyntaxError: Unexpected token export
    at Module._compile (internal/modules/cjs/loader.js:723:23)
    at Object.Module._extensions..js (internal/modules/cjs/loader.js:789:10)
    at Module.load (internal/modules/cjs/loader.js:653:32)
    at tryModuleLoad (internal/modules/cjs/loader.js:593:12)
    at Function.Module._load (internal/modules/cjs/loader.js:585:3)
    at Module.require (internal/modules/cjs/loader.js:692:17)
    at require (internal/modules/cjs/helpers.js:25:18)
    at importCsjOrEsModule (.../hardhat/packages/hardhat-core/internal/core/config/config-loading.js:19:22)
    at Object.loadConfigAndTasks (.../hardhat/packages/hardhat-core/internal/core/config/config-loading.js:48:22)
    at Object.<anonymous> (.../hardhat/packages/hardhat-core/internal/lib/hardhat-lib.js:30:37)
```